pages/[packageName].tsx

```diff
@@ -0,0 +1,49 @@
+import { GetServerSideProps, GetStaticPaths, GetStaticProps, NextPage } from "next";
+import { useRouter } from "next/router";
+import useSwr from "swr";
+import PackageDetails from "../components/PackageDetails";
+
+type PackageDetailsProps = {
+    name: string,
+    version: string,
+    description?: string,
+    depends?: string[],
+    dependant?: string[]
+}
+
+const fetcher = (url: string) => fetch(url).then((res) => res.json())
+
+const PackageDetailsPage: NextPage = () => {
+
+    const router = useRouter()
+    const { data, error } = useSwr(
+        router.query.packageName ? `/api/package-reader/${router.query.packageName}` : null,
+        fetcher
+    )
```

Comment:
Quiz: how would you solve this without relying on `swr`?

Reply:
I would Google? 😄

Reply:
Jokes apart, yes I already so another example which was not using swr at all, but once I saw the one with swr, it looked cleaner so decided on using it.

Reply:
alright, fair enough :) 

Just to be a bit nasty and press you a bit more: 

1. What is the built-in hook that would be required here?
2. How does it work, on a conceptual level?
3. Why is it needed?

I'm pressing on this a bit more, since while being able to find and use libraries is crucial, it's even more crucial to understand how the building blocks of React work. It's annoyingly (and unfortunately) easy to accidentally write code that works 95% of the time, but that 5% bug is very weird to track down.

Reply:
Okay, so I went on and read about the `useEffect` hook (and the useState hook) and the more or less the idea behind Hooks in React.
So I guess `useEffect` hook is the React provided hook that can be used in here instead of `useSwr`.
This hook is used to run additional code after the component has been rendered. This additional code is considered an effect or side effect like for example when fetching data from some network api or manually updating the component DOM after it has completed rendering. But the special characteristic of this hook and I suppose the hooks in general is so that it allows to run use this effects without having to use the React component lifecycle through classes by helping the dev to completely avoid the use of classes and at the same time be able to make use of this React features like state preserving or running side effect code after component is rendered.

I changed the code so it would use this hook along with the useState one instead of the `useSwr`. Suggestions welcomed of course 😄

Reply:
I think you got it yeah!

One thing to keep in mind is proper use of the dependencies array in `useEffect` (the second parameter). I don't remember how [closures](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) work in Java, but as a rule of thumb is to list every variable/constant that you use from outside of the `useEffect` - this way you can be sure that the values inside the `useEffect` have the values that you expect, as the hook is re-ran whenever any of those dependencies change.

Reply:
The closures in Java (called lambada expressions Java 8+) are somewhat similar but probably still not identical. You can use a variable from the outer scope within a closure/lambda but that variable should be final or effectively final in Java meaning it can be assigned only once either at declaration time or before being used within the closure/lambda function. But anyway this "limitation" is easily overcome using objects as you still can assign the object once but you can then change the object properties or instead use some utility classes called AtomicReference, AtomicInteger and so on.

But back to the closure in JavaScript and specifically to the useEffect hook logic, I got the idea on why you have to list those variables within the second argument of the useEffect hook and how the hook internal itself uses those variables to compare them with their previous values from the previous hook execution in order to decide if they have change and if the effect should be executed again.
In this particular case my assessment was that only the `router.query.packageName` is a valid variable to feed the useEffect hook with based on which the hook has to decide to execute its logic or not.